{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bucket 'wgbsfeature' already exists. Cannot create bucket with the same name. Please choose another name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|          | 155/15315 [00:06<07:00, 36.02it/s]"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os \n",
    "import pathlib \n",
    "import glob\n",
    "from datetime import datetime\n",
    "from tqdm import tqdm\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "from elasticsearch import Elasticsearch\n",
    "\n",
    "import minio\n",
    "from minio.error import S3Error\n",
    "from minio.commonconfig import ENABLED\n",
    "from minio.versioningconfig import VersioningConfig\n",
    "\n",
    "from RDSBucket_class import *\n",
    "from data_profiles import *\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "#####----------------------------------------------------------------#####\n",
    "##### preprocessing metadata for MRD and WGS feature dataset.\n",
    "#####----------------------------------------------------------------#####\n",
    "\n",
    "# path_to_project_src = \"/Users/hieunguyen/src/DVC_system\"\n",
    "path_to_project_src = \"/Users/hieunguyen/src/DVC_system\"\n",
    "\n",
    "path_to_save_prep_metadata = os.path.join(path_to_project_src, \"examples/dummy_from_real/prep_metadata\")\n",
    "os.system(\"mkdir -p {}\".format(path_to_save_prep_metadata))\n",
    "\n",
    "feature_description = {\n",
    "    \"CNA\": \"CNA - copy number aberration feature, from WGBS low depth data\",\n",
    "    \"EM\": \"EM - end motif, 4bp\",\n",
    "    \"EM21\":  \"similary to EM, but 21bp end motif\",\n",
    "    \"FLEN\": \"fragment length distribution of all dna fragments\",\n",
    "    \"GWFP_long\": \"normalized number of long fragments, >= 150bp dna fragments\",\n",
    "    \"GWFP_ratio\": \"ratio of short to long fragment, less than 150 to greater than 150bp dna fragments\",\n",
    "    \"GWFP_short\": \"normalized number of short fragments, < 150bp dna fragments\",\n",
    "    \"GWFP_total\": \"normalized number of total fragments\",\n",
    "    \"GWMD\": \"genome wide methylation density\",\n",
    "    \"TMD\": \"targeted methylation density at 450 regions\"   \n",
    "}\n",
    "\n",
    "data_name = \"WGBS_lowdepth\"\n",
    "sub_data_name = \"GW_TM_feature_v0.1\"\n",
    "metadata = pd.read_excel(os.path.join(path_to_project_src, \"/ECD_metadata/metadata_cfDNA_lowpdepth.xlsx\"))\n",
    "\n",
    "for feature_name in feature_description.keys():\n",
    "    all_files = [item for item in pathlib.Path(\"{}/examples/dummy_from_real/{}/{}/{}\".format(path_to_project_src, data_name, sub_data_name, feature_name)).glob(\"*\")\n",
    "                if \".DS\" not in item.name]  \n",
    "    \n",
    "    if os.path.isfile(os.path.join(path_to_save_prep_metadata, \"{}.{}.{}.metadata.csv\".format(data_name, sub_data_name, feature_name))) == False:\n",
    "        maindf = pd.DataFrame(data = [str(item) for item in all_files], columns=[\"path\"])\n",
    "        maindf[\"FileName\"] = maindf[\"path\"].apply(lambda x: \"{}_{}\".format(feature_name, x.split(\"/\")[-1]))\n",
    "        maindf[\"FeatureName\"] = feature_name\n",
    "        maindf[\"Labcode\"] = maindf[\"FileName\"].apply(lambda x: x.replace(\".csv\", \"\").replace(\"{}_\".format(feature_name), \"\"))\n",
    "        maindf = maindf[maindf[\"Labcode\"].isin(metadata.SampleID.unique())]\n",
    "        maindf[\"SequencingID\"] = maindf[\"Labcode\"]\n",
    "        maindf[\"Description\"] = feature_description[feature_name]\n",
    "        maindf[\"FeatureVersion\"] = \"v0.1\"\n",
    "        maindf[\"Date\"] = datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "        maindf[\"Pipeline\"] = \"ECD_WGBS_bsalign\"\n",
    "        maindf[\"Pipeline_params\"] = \"default\"\n",
    "        maindf[\"Pipeline_repo\"] = \"gitlab\"\n",
    "        maindf[\"Feature_Pipeline\"] = \"WCD_WGBS_bsalign_GW_TM_feature_v0.1\"\n",
    "        maindf[\"Feature_Pipeline_params\"] = \"default\"\n",
    "        maindf[\"Feature_Pipeline_repo\"] = \"gitlab\"\n",
    "        maindf[\"Ref_genome\"] = \"hg19\"\n",
    "        maindf[\"Input\"] = \"bsalign files\"\n",
    "        maindf[\"Project\"] = \"ECD_WGBS\"\n",
    "        maindf[\"Sub_project\"] = \"lowdepth_data\"\n",
    "\n",
    "        maindf[\"label1\"] = maindf[\"Labcode\"].apply(lambda x: metadata[metadata[\"SampleID\"] == x][\"Label\"].values[0] if x in metadata[\"SampleID\"].values else \"not available\")\n",
    "        maindf[\"label2\"] = maindf[\"Labcode\"].apply(lambda x: metadata[metadata[\"SampleID\"] == x][\"Run_TM\"].values[0] if x in metadata[\"SampleID\"].values else \"not available\")\n",
    "        maindf[\"label3\"] = maindf[\"Labcode\"].apply(lambda x: metadata[metadata[\"SampleID\"] == x][\"Run_GW\"].values[0] if x in metadata[\"SampleID\"].values else \"not available\")\n",
    "\n",
    "        maindf[\"label4\"] = \"not available\"\n",
    "\n",
    "        maindf[\"Note\"] = \"not available\"\n",
    "        maindf[\"member\"] = \"hieunguyen,hieutran\"\n",
    "\n",
    "        maindf.to_csv(os.path.join(path_to_save_prep_metadata, \"{}.{}.{}.metadata.csv\".format(data_name, sub_data_name, feature_name)), index = False)\n",
    "    else:\n",
    "        maindf = pd.read_csv(os.path.join(path_to_save_prep_metadata, \"{}.{}.{}.metadata.csv\".format(data_name, sub_data_name, feature_name)))\n",
    "        print(\"The metadata has been already generated\")\n",
    "    maindf = maindf.set_index(\"path\")\n",
    "    input_dict = maindf.to_dict(orient = \"index\") # the input metadata is ready to be added to the database elasticsearch\n",
    "\n",
    "    ##### Credential\n",
    "    minio_credential = os.path.join(path_to_project_src, \"credentials.macstudio.json\")\n",
    "    es_credential = os.path.join(path_to_project_src, \"es_credential.json\")\n",
    "\n",
    "    bamBucket = RDSBucket( \n",
    "                            minio_credential = minio_credential, \n",
    "                            bucketName = \"wgbsfeature\",\n",
    "                            PROFILE_NAME = \"wgbsfeature\",\n",
    "                            DATA_PROFILES = DATA_PROFILES,\n",
    "                            es_credential = es_credential, \n",
    "                            versioning = True, \n",
    "                            verbose = False)\n",
    "    bamBucket.initBucket()\n",
    "\n",
    "    for path in tqdm(input_dict.keys()):\n",
    "        file_metadata = input_dict[path]\n",
    "        bamBucket.upload_file_to_bucket(path_to_file = path, \n",
    "                                        object_name= file_metadata[\"FileName\"], \n",
    "                                        file_metadata = file_metadata)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "minio_dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
